File: backend/src/services/rental.service.js
const mongoose = require('mongoose');
const { Rental, Item, Conversation } = require('../models');
const ApiError = require('../utils/apiError');
const notificationService = require('./notification.service');

/**
 * Creates a new rental request.
 * Validates that the renter is not the owner and the item is available.
 * Also creates an associated conversation for the rental.
 * Input: { itemId, startDate, endDate, message }, renterId (string)
 * Output: populated rental object
 */
async function createRental({ itemId, startDate, endDate, message, offerPrice }, renterId) {
  const item = await Item.findById(itemId).populate('owner', 'name');
  if (!item) {
    throw ApiError.notFound('Item not found');
  }

  const start = new Date(startDate);
  const end = new Date(endDate);
  
  if (isNaN(start.getTime()) || isNaN(end.getTime())) {
    throw ApiError.badRequest('Invalid date format');
  }
  if (end <= start) {
    throw ApiError.badRequest('End date must be after start date');
  }

  // Check if item is available for these specific dates
  const overlappingRental = await Rental.findOne({
    item: itemId,
    status: { $in: ['approved', 'active'] },
    $or: [
      { startDate: { $lte: start }, endDate: { $gte: start } },
      { startDate: { $lte: end }, endDate: { $gte: end } },
      { startDate: { $gte: start }, endDate: { $lte: end } }
    ]
  });

  if (overlappingRental) {
    throw ApiError.badRequest('Item is already booked for the selected dates');
  }

  if (item.owner.toString() === renterId) {
    throw ApiError.badRequest('You cannot rent your own item');
  }

  const days = Math.ceil((end - start) / (1000 * 60 * 60 * 24));
  const totalPrice = days * item.pricePerDay;

  const session = await mongoose.startSession();
  session.startTransaction();

  try {
    const rental = await Rental.create(
      [
        {
          item: itemId,
          renter: renterId,
          owner: item.owner,
          startDate: start,
          endDate: end,
          totalPrice,
          message: message || '',
          offerPrice: (offerPrice !== undefined && offerPrice !== null) ? Number(offerPrice) : null,
        },
      ],
      { session }
    );

    // Create a conversation linked to this rental
    await Conversation.create(
      [
        {
          rental: rental[0]._id,
          participants: [renterId, item.owner._id.toString()],
        },
      ],
      { session }
    );

    // Notify the owner of the new request
    await notificationService.createNotification({
      recipient: item.owner._id,
      sender: renterId,
      type: 'rental_request',
      title: 'New Rental Request',
      message: `Someone wants to rent your ${item.title}!`,
      link: `/rentals/${rental[0]._id}`,
    }, { session });

    await session.commitTransaction();

    const populatedRental = await rental[0].populate([
      { path: 'item', select: 'title images pricePerDay' },
      { path: 'renter', select: 'name email avatar' },
      { path: 'owner', select: 'name email avatar' },
    ]);
    return populatedRental;
  } catch (error) {
    await session.abortTransaction();
    throw error;
  } finally {
    session.endSession();
  }
}

/**
 * Fetches all rentals where the user is the renter.
 * Input: userId (string)
 * Output: array of populated rental objects
 */
async function getMyRentals(userId) {
  return Rental.find({ renter: userId })
    .populate('item', 'title images pricePerDay category')
    .populate('owner', 'name email avatar campus')
    .sort({ createdAt: -1 });
}

/**
 * Fetches all rental requests received by the item owner.
 * Input: userId (string)
 * Output: array of populated rental objects
 */
async function getReceivedRequests(userId) {
  return Rental.find({ owner: userId })
    .populate('item', 'title images pricePerDay category')
    .populate('renter', 'name email avatar campus')
    .sort({ createdAt: -1 });
}

/**
 * Updates the status of a rental request.
 * Only the owner can approve/reject; either party can cancel.
 * Input: rentalId (string), status (string), userId (string)
 * Output: updated rental object
 */
async function updateRentalStatus(rentalId, status, userId) {
  const rental = await Rental.findById(rentalId).populate('item');
  if (!rental) {
    throw ApiError.notFound('Rental not found');
  }

  const isOwner = rental.owner.toString() === userId;
  const isRenter = rental.renter.toString() === userId;

  if (!isOwner && !isRenter) {
    throw ApiError.forbidden('You are not part of this rental');
  }

  // Only owner can approve or reject
  if (['approved', 'rejected'].includes(status) && !isOwner) {
    throw ApiError.forbidden('Only the item owner can approve or reject');
  }

  // Only allow cancellation by either party
  if (status === 'cancelled' && !isOwner && !isRenter) {
    throw ApiError.forbidden('You cannot cancel this rental');
  }

  // Validate status transitions
  const validTransitions = {
    pending: ['approved', 'rejected', 'cancelled'],
    approved: ['active', 'cancelled'],
    active: ['completed', 'cancelled'],
  };

  const oldStatus = rental.status;

  if (!validTransitions[oldStatus]?.includes(status)) {
    throw ApiError.badRequest(`Cannot change status from '${oldStatus}' to '${status}'`);
  }

  const session = await mongoose.startSession();
  session.startTransaction();

  try {
    rental.status = status;
    await rental.save({ session });

    if (status === 'approved') {
      // Check for overlap one last time before committing
      const overlappingRental = await Rental.findOne({
        _id: { $ne: rental._id },
        item: rental.item._id,
        status: { $in: ['approved', 'active'] },
        $or: [
          { startDate: { $lte: rental.startDate }, endDate: { $gte: rental.startDate } },
          { startDate: { $lte: rental.endDate }, endDate: { $gte: rental.endDate } },
          { startDate: { $gte: rental.startDate }, endDate: { $lte: rental.endDate } }
        ]
      }).session(session);

      if (overlappingRental) {
        throw ApiError.badRequest('This item is already booked for overlapping dates.');
      }

      // ONLY cancel pending rentals that OVERLAP with this one
      await Rental.updateMany(
        { 
          item: rental.item._id, 
          _id: { $ne: rental._id }, 
          status: 'pending',
          $or: [
            { startDate: { $lte: rental.startDate }, endDate: { $gte: rental.startDate } },
            { startDate: { $lte: rental.endDate }, endDate: { $gte: rental.endDate } },
            { startDate: { $gte: rental.startDate }, endDate: { $lte: rental.endDate } }
          ]
        },
        { 
          status: 'cancelled',
          message: 'This request was cancelled because the item was booked for overlapping dates.' 
        },
        { session }
      );

      // Notify the renter that their request was approved
      await notificationService.createNotification({
        recipient: rental.renter,
        sender: userId,
        type: 'rental_status',
        title: 'Rental Approved!',
        message: `Your request for ${rental.item.title} has been approved.`,
        link: `/rentals/${rental._id}`,
      }, { session });

    } else if (status === 'rejected') {
      // Notify the renter of rejection
      await notificationService.createNotification({
        recipient: rental.renter,
        sender: userId,
        type: 'rental_status',
        title: 'Request Rejected',
        message: `Your request for ${rental.item.title} was declined.`,
        link: `/rentals/${rental._id}`,
      }, { session });

    } else if (status === 'active') {
      // Notify the renter that rental is now active
      await notificationService.createNotification({
        recipient: rental.renter,
        sender: userId,
        type: 'rental_status',
        title: 'Rental Started',
        message: `You are now renting the ${rental.item.title}.`,
        link: `/rentals/${rental._id}`,
      }, { session });

    } else if (status === 'completed') {
      // Notify both parties to leave reviews
      await notificationService.createNotification({
        recipient: rental.renter,
        sender: userId,
        type: 'rental_status',
        title: 'Rental Completed',
        message: `How was your experience? Please rate the item and the owner!`,
        link: `/rentals/${rental._id}`,
      }, { session });

      await notificationService.createNotification({
        recipient: rental.owner,
        sender: userId,
        type: 'rental_status',
        title: 'Rental Completed',
        message: `How was your experience? Please rate the renter!`,
        link: `/rentals/${rental._id}`,
      }, { session });

    } else if (status === 'cancelled') {
      // Notify the other party of cancellation
      const otherPartyId = isOwner ? rental.renter : rental.owner;
      await notificationService.createNotification({
        recipient: otherPartyId,
        sender: userId,
        type: 'rental_status',
        title: 'Rental Cancelled',
        message: `The rental for ${rental.item.title} has been cancelled.`,
        link: `/rentals/${rental._id}`,
      }, { session });
    }

    await session.commitTransaction();

    const populatedRental = await rental.populate([
      { path: 'item', select: 'title images pricePerDay' },
      { path: 'renter', select: 'name email avatar' },
      { path: 'owner', select: 'name email avatar' },
    ]);
    return populatedRental;
  } catch (error) {
    await session.abortTransaction();
    throw error;
  } finally {
    session.endSession();
  }
}

/**
 * Fetches a single rental by ID with full details.
 * Input: rentalId (string), userId (string)
 * Output: populated rental object
 */
async function getRentalById(rentalId, userId) {
  const rental = await Rental.findById(rentalId)
    .populate('item', 'title images pricePerDay category description')
    .populate('renter', 'name email avatar campus')
    .populate('owner', 'name email avatar campus');

  if (!rental) {
    throw ApiError.notFound('Rental not found');
  }

  const isOwner = rental.owner._id.toString() === userId;
  const isRenter = rental.renter._id.toString() === userId;

  if (!isOwner && !isRenter) {
    throw ApiError.forbidden('You are not part of this rental');
  }

  return rental;
}

module.exports = {
  createRental,
  getMyRentals,
  getReceivedRequests,
  updateRentalStatus,
  getRentalById,
};

File: backend/src/main/java/com/rentify/rental/RentalService.java
package com.rentify.rental;

import com.rentify.conversation.Conversation;
import com.rentify.conversation.ConversationRepository;
import com.rentify.exception.BadRequestException;
import com.rentify.exception.ForbiddenException;
import com.rentify.exception.NotFoundException;
import com.rentify.item.Item;
import com.rentify.item.ItemRepository;
import com.rentify.notification.NotificationService;
import com.rentify.notification.NotificationType;
import com.rentify.rental.dto.CreateRentalRequest;
import com.rentify.rental.dto.RentalResponse;
import com.rentify.rental.dto.UpdateRentalStatusRequest;
import com.rentify.user.User;
import com.rentify.user.UserRepository;
import com.rentify.user.UserRole;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.List;

@Service
public class RentalService {

    private final RentalRepository rentalRepository;
    private final ItemRepository itemRepository;
    private final UserRepository userRepository;
    private final NotificationService notificationService;
    private final ConversationRepository conversationRepository;

    public RentalService(
            RentalRepository rentalRepository,
            ItemRepository itemRepository,
            UserRepository userRepository,
            NotificationService notificationService,
            ConversationRepository conversationRepository
    ) {
        this.rentalRepository = rentalRepository;
        this.itemRepository = itemRepository;
        this.userRepository = userRepository;
        this.notificationService = notificationService;
        this.conversationRepository = conversationRepository;
    }

    @Transactional
    public RentalResponse createRental(Long renterId, CreateRentalRequest request) {
        if (request.startDate().isBefore(LocalDate.now())) {
            throw new BadRequestException("Start date cannot be in the past");
        }
        if (request.endDate().isBefore(request.startDate())) {
            throw new BadRequestException("End date cannot be before start date");
        }

        User renter = userRepository.findById(renterId)
                .orElseThrow(() -> new NotFoundException("User not found"));

        Item item = itemRepository.findById(request.itemId())
                .orElseThrow(() -> new NotFoundException("Item not found"));

        if (item.getOwner().getId().equals(renterId)) {
            throw new BadRequestException("You cannot rent your own item");
        }

        if (!item.isAvailable()) {
            throw new BadRequestException("Item is currently not available for rent");
        }

        boolean hasOverlap = rentalRepository.hasOverlappingRentals(
                item.getId(),
                request.startDate(),
                request.endDate(),
                List.of(RentalStatus.APPROVED, RentalStatus.ACTIVE)
        );
        if (hasOverlap) {
            throw new BadRequestException("Item is not available for the selected dates");
        }

        long days = ChronoUnit.DAYS.between(request.startDate(), request.endDate()) + 1;
        BigDecimal totalPrice = item.getPricePerDay().multiply(BigDecimal.valueOf(days));

        Rental rental = new Rental();
        rental.setRenter(renter);
        rental.setItem(item);
        rental.setOwner(item.getOwner());
        rental.setStartDate(request.startDate());
        rental.setEndDate(request.endDate());
        rental.setTotalPrice(totalPrice);
        rental.setStatus(RentalStatus.PENDING);
        rental.setMessage(request.message() != null ? request.message().trim() : "");

        Rental savedRental = rentalRepository.save(rental);

        Conversation conversation = new Conversation(savedRental, renter, item.getOwner());
        conversationRepository.save(conversation);

        notificationService.createNotification(
                item.getOwner(),
                renter,
                NotificationType.RENTAL_REQUEST,
                "New Rental Request",
                renter.getName() + " requested to rent " + item.getTitle(),
                "/rentals/" + savedRental.getId()
        );

        return RentalResponse.fromEntity(savedRental);
    }

    @Transactional(readOnly = true)
    public List<RentalResponse> getMyRentals(Long renterId) {
        List<Rental> rentals = rentalRepository.findByRenterIdOrderByCreatedAtDesc(renterId);
        return rentals.stream()
                .map(RentalResponse::fromEntity)
                .toList();
    }

    @Transactional(readOnly = true)
    public List<RentalResponse> getReceivedRentals(Long ownerId) {
        List<Rental> rentals = rentalRepository.findByOwnerIdOrderByCreatedAtDesc(ownerId);
        return rentals.stream()
                .map(RentalResponse::fromEntity)
                .toList();
    }

    @Transactional(readOnly = true)
    public RentalResponse getRentalById(Long userId, Long rentalId) {
        Rental rental = rentalRepository.findById(rentalId)
                .orElseThrow(() -> new NotFoundException("Rental not found"));

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new NotFoundException("User not found"));

        boolean isAuthorized = user.getRole() == UserRole.ADMIN
                || rental.getRenter().getId().equals(userId)
                || rental.getOwner().getId().equals(userId);

        if (!isAuthorized) {
            throw new ForbiddenException("You are not authorized to view this rental");
        }

        return RentalResponse.fromEntity(rental);
    }

    @Transactional
    public RentalResponse updateRentalStatus(Long userId, Long rentalId, UpdateRentalStatusRequest request) {
        Rental rental = rentalRepository.findById(rentalId)
                .orElseThrow(() -> new NotFoundException("Rental not found"));

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new NotFoundException("User not found"));

        RentalStatus currentStatus = rental.getStatus();
        RentalStatus newStatus = request.status();

        switch (newStatus) {
            case APPROVED -> {
                if (!rental.getOwner().getId().equals(userId)) {
                    throw new ForbiddenException("Only the item owner can approve rental requests");
                }
                if (currentStatus != RentalStatus.PENDING) {
                    throw new BadRequestException("Can only approve pending rental requests");
                }
                boolean hasOverlap = rentalRepository.hasOverlappingRentals(
                        rental.getItem().getId(),
                        rental.getStartDate(),
                        rental.getEndDate(),
                        List.of(RentalStatus.APPROVED, RentalStatus.ACTIVE)
                );
                if (hasOverlap) {
                    throw new BadRequestException("Cannot approve: another rental already overlaps with these dates");
                }
                rental.setStatus(RentalStatus.APPROVED);
                notificationService.createNotification(
                        rental.getRenter(),
                        user,
                        NotificationType.RENTAL_STATUS,
                        "Rental Request Approved",
                        "Your request to rent " + rental.getItem().getTitle() + " has been approved!",
                        "/rentals/" + rental.getId()
                );
            }
            case REJECTED -> {
                if (!rental.getOwner().getId().equals(userId)) {
                    throw new ForbiddenException("Only the item owner can reject rental requests");
                }
                if (currentStatus != RentalStatus.PENDING) {
                    throw new BadRequestException("Can only reject pending rental requests");
                }
                rental.setStatus(RentalStatus.REJECTED);
                notificationService.createNotification(
                        rental.getRenter(),
                        user,
                        NotificationType.RENTAL_STATUS,
                        "Rental Request Declined",
                        "Your request to rent " + rental.getItem().getTitle() + " was declined.",
                        "/rentals/" + rental.getId()
                );
            }
            case CANCELLED -> {
                boolean isParticipant = rental.getRenter().getId().equals(userId) || rental.getOwner().getId().equals(userId);
                if (!isParticipant) {
                    throw new ForbiddenException("You are not authorized to cancel this rental");
                }
                if (currentStatus != RentalStatus.PENDING && currentStatus != RentalStatus.APPROVED) {
                    throw new BadRequestException("Cannot cancel a rental that is " + currentStatus.getValue());
                }
                rental.setStatus(RentalStatus.CANCELLED);
                User recipient = rental.getRenter().getId().equals(userId) ? rental.getOwner() : rental.getRenter();
                String message = rental.getRenter().getId().equals(userId)
                        ? rental.getRenter().getName() + " cancelled their rental request for " + rental.getItem().getTitle()
                        : "The rental for " + rental.getItem().getTitle() + " was cancelled by the owner.";

                notificationService.createNotification(
                        recipient,
                        user,
                        NotificationType.RENTAL_STATUS,
                        "Rental Cancelled",
                        message,
                        "/rentals/" + rental.getId()
                );
            }
            case ACTIVE -> {
                boolean isParticipant = rental.getRenter().getId().equals(userId) || rental.getOwner().getId().equals(userId);
                if (!isParticipant) {
                    throw new ForbiddenException("You are not authorized to start this rental");
                }
                if (currentStatus != RentalStatus.APPROVED) {
                    throw new BadRequestException("Can only start an approved rental");
                }
                rental.setStatus(RentalStatus.ACTIVE);
                User recipient = rental.getRenter().getId().equals(userId) ? rental.getOwner() : rental.getRenter();
                notificationService.createNotification(
                        recipient,
                        user,
                        NotificationType.RENTAL_STATUS,
                        "Rental Started",
                        "The rental period for " + rental.getItem().getTitle() + " is now active.",
                        "/rentals/" + rental.getId()
                );
            }
            case COMPLETED -> {
                if (!rental.getOwner().getId().equals(userId)) {
                    throw new ForbiddenException("Only the item owner can mark the rental as completed");
                }
                if (currentStatus != RentalStatus.ACTIVE) {
                    throw new BadRequestException("Can only complete an active rental");
                }
                rental.setStatus(RentalStatus.COMPLETED);
                notificationService.createNotification(
                        rental.getRenter(),
                        user,
                        NotificationType.RENTAL_STATUS,
                        "Rental Completed",
                        "The rental for " + rental.getItem().getTitle() + " has been marked as completed. Please leave a review!",
                        "/rentals/" + rental.getId()
                );
            }
            default -> throw new BadRequestException("Invalid status transition");
        }

        if (request.message() != null && !request.message().isBlank()) {
            rental.setMessage(request.message().trim());
        }

        Rental updatedRental = rentalRepository.save(rental);
        return RentalResponse.fromEntity(updatedRental);
    }
}
